File: src/src/main/java/br/com/blz/testjava/services/ProductService.java
package br.com.blz.testjava.services;

import br.com.blz.testjava.exception.ProductAlreadyExistsException;
import br.com.blz.testjava.exception.RecordNotFoundException;
import br.com.blz.testjava.models.Product;
import br.com.blz.testjava.repositories.ProductRepository;
import lombok.AllArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;


@AllArgsConstructor
@Service
public class ProductService {

    private ProductRepository productRepository;

    public Product saveProduct(Product product) {
        if(productRepository.existsBySku(product.getSku())){
            throw new ProductAlreadyExistsException("Product Already Exists");
        }

        return productRepository.save(product);
    }

    public Product findBySku(Long sku){
        if(!productRepository.existsBySku(sku)){
            throw new RecordNotFoundException("Product Not Found");
        }
        return productRepository.findBySku(sku);
    }

    public void deleteProduct(Long sku){
        if(!productRepository.existsBySku(sku)){
            throw new RecordNotFoundException("Product Not Found");
        }

        Product product = productRepository.findBySku(sku);
        productRepository.delete(product);
    }
}

File: src/main/java/br/com/blz/testjava/service/ProductService.kt
package br.com.blz.testjava.service

import br.com.blz.testjava.model.Product
import br.com.blz.testjava.repository.ProductRepository
import org.springframework.stereotype.Service

@Service
class ProductService (private val productRepository: ProductRepository){

  fun save(entity: Product) = productRepository.save(entity)
  fun findBySku(sku: Long) = productRepository.findBySku(sku)
  fun findAll()  = productRepository.findAll()
  fun delete(sku: Long) {
    val product = findBySku(sku)

    productRepository.delete(product)
  }
}
